plugins/processors/t128_rate/t128_rate.go

```diff
@@ -0,0 +1,164 @@
+package rename
+
+import (
+	"fmt"
+	"math"
+	"sort"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+const sampleConfig = `
+	## If more than this amount of time passes between data points, the
+	## previous value will be considered old and the rate will be recalculated
+	## as if it hadn't been seen before. A zero expiration means never expire.
+	# expiration = "0s"
+
+	## For the fields who's key/value pairs don't match, should the original
+	## field be removed?
+	# remove-original = true
+
+[processors.t128_rate.fields]
+	## Replace fields with their rates, renaming them if indicated
+	# "/rate/metric" = "/total/metric"
+	# "/inline/replace" = "/inline/replace"
+`
+
+type T128Rate struct {
+	Fields         map[string]string `toml:"fields"`
+	Expiration     internal.Duration `toml:"expiration"`
+	RemoveOriginal bool              `toml:"remove-original"`
+
+	Log telegraf.Logger `toml:"-"`
+
+	targetFields map[string]target
+	cache        map[uint64]map[string]observedValue
+}
+
+type target struct {
+	key           string
+	matchesSource bool
+}
+
+type observedValue struct {
+	value   float64
+	expires time.Time
+}
+
+func (r *T128Rate) SampleConfig() string {
+	return sampleConfig
+}
+
+func (r *T128Rate) Description() string {
+	return "Compute the rate of fields that pass through this filter."
+}
+
+func (r *T128Rate) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	for _, point := range in {
+		seriesHash := point.HashID()
+
+		removeFields := make([]string, 0)
+
+		for _, field := range point.FieldList() {
+			target, ok := r.targetFields[field.Key]
+			if !ok {
+				continue
+			}
+
+			currentValue, converted := convert(field.Value)
+			if !converted {
+				r.Log.Warnf("Failed to convert field '%s' to float for a rate calculation. The rate computation will be skipped.")
+				continue
+			}
+
+			cacheFields, metricIsCached := r.cache[seriesHash]
+			if !metricIsCached {
+				r.cache[seriesHash] = make(map[string]observedValue, 0)
```

Comment:
The `zero` is the initial value? Shouldn't it use the value you just got for this metric in this function?

Reply:
This is just creating the map for this series and `zero` is the initial size of the map. The value for the particular metric ID is set on line 94.

A series is the unique combination of measurement and tags represented by the `seriesHash`. For the series, I'm creating an empty map to hold the values for each metric ID that appears with those tags.